### PR TITLE
Add to Pythia8Interface the plugin that uses HepMC3

### PR DESCRIPTION
**Pull request by mkirsano**
#### PR description:

The new plugin produces HepMC3 products. For the moment only GEN step can be run like this.

#### PR validation:

Local tests using the new configurations in /test

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by mkirsano** [comment]
please test

**Comment by mkirsano** [comment]
please test

**Comment by Saptaparna** [comment]
+1

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8HepMC3Hadronizer.cc [hunk 1]
```diff
@@ -0,0 +1,1089 @@
...
+    throw edm::Exception(edm::errors::Configuration, "Pythia8Interface")
+        << " Obsolete parameter: VinciaPlugin \n Please use the parameter PartonShowers:model instead \n";
+  }
+  if (params.exists("DirePlugin")) {
+    throw edm::Exception(edm::errors::Configuration, "Pythia8Interface")
+        << " Obsolete parameter: DirePlugin \n Please use the parameter PartonShowers:model instead \n";
+  }
+}
+
+Pythia8HepMC3Hadronizer::~Pythia8HepMC3Hadronizer() {}
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8HepMC3Hadronizer.cc:342]:
```suggestion
```

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8HepMC3Hadronizer.cc [hunk 2]
```diff
@@ -0,0 +1,1089 @@
...
+  for (auto p : (event3().get())->particles()) {
+    NPartsAfterDecays++;
+  }
+
+  if (NPartsAfterDecays == NPartsBeforeDecays)
+    return true;
+
+  bool result = true;
+
+  std::cout << "residualdecay in action " << NPartsBeforeDecays << " " << NPartsAfterDecays << std::endl;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8HepMC3Hadronizer.cc:912]: Please do not use thread unsafe `cout`. Use LogSomething instead (here and below, unless if screened by an appropriate verbosity flag, which I assume is only used for debug)

File: GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+#include "GeneratorInterface/Pythia8Interface/interface/Py8HMC3InterfaceBase.h"
+#include "GeneratorInterface/Pythia8Interface/interface/SLHAReaderBase.h"
+
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include <filesystem>
+
+// EvtGen plugin
+//
+//#include "Pythia8Plugins/EvtGen.h"
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc:13]: Remove if not needed

File: GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        evtgenPdlFile = pdt.fullPath();
+      }
+
+      if (ps.exists("evtgenUserFile")) {
+        std::vector<std::string> user_decays = ps.getParameter<std::vector<std::string> >("evtgenUserFile");
+        for (unsigned int i = 0; i < user_decays.size(); i++) {
+          edm::FileInPath user_decay(user_decays.at(i));
+          evtgenUserFiles.push_back(user_decay.fullPath());
+        }
+        //evtgenUserFiles = ps.getParameter< std::vector<std::string> >("evtgenUserFile");
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc:53]: Please clean up of whatever not needed

File: GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+    slhafile_ = tempslhaname;
+
+    fMasterGen->settings.mode("SLHA:readFrom", 2);
+    fMasterGen->settings.word("SLHA:file", slhafile_);
+  }
+
+  bool Py8HMC3InterfaceBase::declareStableParticles(const std::vector<int>& pdgIds) {
+    for (size_t i = 0; i < pdgIds.size(); i++) {
+      // FIXME: need to double check if PID's are the same in Py6 & Py8,
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc:187]: Was this checked/fixed already?

File: GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+using namespace Pythia8;
+
+namespace gen {
+
+  Py8HMC3InterfaceBase::Py8HMC3InterfaceBase(edm::ParameterSet const& ps)
+      : BaseHadronizer(ps), useEvtGen(false), evtgenDecays(nullptr) {
+    fParameters = ps;
+
+    pythiaPylistVerbosity = ps.getUntrackedParameter<int>("pythiaPylistVerbosity", 0);
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/src/Py8HMC3InterfaceBase.cc:23]: If these parameters get initialized in the class initializing list, they can be better declared as `const` in the header

**Comment by mkirsano** [comment]
please test

**Comment by mkirsano** [comment]
+1

**Comment by mkirsano** [comment]
After asking around I removed the comment that seems obsolete.

**Comment by mkirsano** [comment]
Note that the code is not changed

**Comment by mkirsano** [comment]
please test

**Comment by mkirsano** [comment]
please test

**Pull request merged**